src/core/server/http/base_path_proxy_server.ts

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { RouteOptions, Server } from 'hapi-latest';
+import { Agent as HttpsAgent, ServerOptions as TlsOptions } from 'https';
+import { sample } from 'lodash';
+import { ByteSizeValue } from '../config/schema';
+import { DevConfig } from '../dev';
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+const alphabet = 'abcdefghijklmnopqrztuvwxyz'.split('');
+
+export interface BasePathProxyServerOptions {
+  httpConfig: HttpConfig;
+  devConfig: DevConfig;
+  isKibanaPath: (path: string) => boolean;
+  blockUntil?: () => Promise<void>;
+}
+
+export class BasePathProxyServer {
+  private server?: Server;
+  private httpsAgent?: HttpsAgent;
+
+  get basePath() {
+    return this.options.httpConfig.basePath;
+  }
+
+  get targetPort() {
+    return this.options.devConfig.basePathProxyTargetPort;
+  }
+
+  constructor(
+    private readonly log: Logger,
+    private readonly options: BasePathProxyServerOptions
+  ) {
+    const ONE_GIGABYTE = 1024 * 1024 * 1024;
+    options.httpConfig.maxPayload = new ByteSizeValue(ONE_GIGABYTE);
+
+    if (!options.httpConfig.basePath) {
+      options.httpConfig.basePath = `/${sample(alphabet, 3).join('')}`;
+    }
+  }
+
+  public async start() {
```

Comment:
**nit:** up to you, of course, but wondering if you think splitting `start` into smaller pieces would make this easier to read? maybe each `server.route(...)` can be its own method, maybe something like a `setupRoutes` for them all.

Reply:
Yep, I think having `setupRoutes` is a good idea, thanks!